cpp/examples/arrow/join_example.cc

```diff
@@ -0,0 +1,194 @@
+  std::shared_ptr<arrow::io::InputStream> input;
+  std::string csv_data = GetDataAsCsvString(relation);
+  std::cout << "CSV DATA : " << relation << std::endl;
+  std::cout << csv_data << std::endl;
+  arrow::util::string_view sv = csv_data;
+  input = std::make_shared<arrow::io::BufferReader>(sv);
+  auto read_options = arrow::csv::ReadOptions::Defaults();
+  auto parse_options = arrow::csv::ParseOptions::Defaults();
+  auto convert_options = arrow::csv::ConvertOptions::Defaults();
+
+  // Instantiate TableReader from input stream and options
+  ARROW_ASSIGN_OR_RAISE(std::shared_ptr<arrow::csv::TableReader> table_reader,
+                        arrow::csv::TableReader::Make(io_context, input, read_options,
+                                                      parse_options, convert_options));
+
+  std::shared_ptr<arrow::csv::TableReader> reader = table_reader;
+
+  // Read table from CSV file
+  ARROW_ASSIGN_OR_RAISE(auto maybe_table, reader->Read());
+  auto ds = std::make_shared<arrow::dataset::InMemoryDataset>(maybe_table);
+  arrow::Result<std::shared_ptr<arrow::dataset::InMemoryDataset>> result(std::move(ds));
+  return result;
+}
+
+cp::Expression Materialize(std::vector<std::string> names) {
+  std::vector<cp::Expression> exprs;
+  for (const auto& name : names) {
+    exprs.push_back(cp::field_ref(name));
+  }
+
+  return cp::project(exprs, names);
+}
+
+arrow::Status DoHashJoin() {
+  cp::ExecContext exec_context(arrow::default_memory_pool(),
+                               ::arrow::internal::GetCpuThreadPool());
+
+  arrow::dataset::internal::Initialize();
+
+  ARROW_ASSIGN_OR_RAISE(std::shared_ptr<cp::ExecPlan> plan,
+                        cp::ExecPlan::Make(&exec_context));
+
+  arrow::AsyncGenerator<arrow::util::optional<cp::ExecBatch>> sink_gen;
+
+  cp::ExecNode* left_source;
+  cp::ExecNode* right_source;
+
+  ARROW_ASSIGN_OR_RAISE(auto l_dataset, CreateDataSetFromCSVData("l"));
+  ARROW_ASSIGN_OR_RAISE(auto r_dataset, CreateDataSetFromCSVData("r"));
+
+  auto l_options = std::make_shared<arrow::dataset::ScanOptions>();
+  l_options->projection = Materialize({});  // create empty projection
+
+  auto r_options = std::make_shared<arrow::dataset::ScanOptions>();
+  r_options->projection = Materialize({});  // create empty projection
+
+  // construct the scan node
+  auto l_scan_node_options = arrow::dataset::ScanNodeOptions{l_dataset, l_options};
+  auto r_scan_node_options = arrow::dataset::ScanNodeOptions{r_dataset, r_options};
+
+  ARROW_ASSIGN_OR_RAISE(left_source,
+                        cp::MakeExecNode("scan", plan.get(), {}, l_scan_node_options));
+  ARROW_ASSIGN_OR_RAISE(right_source,
+                        cp::MakeExecNode("scan", plan.get(), {}, r_scan_node_options));
+
+  arrow::compute::HashJoinNodeOptions join_opts{
+      arrow::compute::JoinType::INNER,
+      /*left_keys=*/{"lkey"},
+      /*right_keys=*/{"rkey"},         arrow::compute::literal(true), "_l", "_r"};
```

Comment:
```suggestion
      /*in_left_keys=*/{"lkey"},
      /*in_right_keys=*/{"rkey"},         arrow::compute::literal(true), "_l", "_r"};
```
Also, if you're going to specify the names for these parameters you should probably specify the names for the rest.  None of them are intuitive.